core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -352,6 +353,21 @@ class TransactionCoordinator(brokerId: Int,
                            producerEpoch: Short,
                            txnMarkerResult: TransactionResult,
                            responseCallback: EndTxnCallback): Unit = {
+    endTransaction(transactionalId,
+      producerId,
+      producerEpoch,
+      txnMarkerResult,
+      // Strict equality is enforced on the client side requests, as they shouldn't bump the producer epoch.
+      (producerEpoch, txnMetadata) => producerEpoch != txnMetadata.producerEpoch,
```

Comment:
This is a little awkward. I think the `isFromClient` flag was easier to understand and avoided the need to duplicate the check in a couple places.